### PR TITLE
Credentials page/Self cred update flow UI improvements

### PR DESCRIPTION
**Pull request by weijiangan**
# Change summary
Didn't realize #2995 is being done already, but it was not all in vain. I tried to fit the cred update flow fit into the normal pages (without affecting the original reset screen) and made the sidebar functional. In my head we can repurpose the cred update flow... to also be the credential status page? I was looking at GitHub's settings page and it didn't look too far apart. Would love to know what you think @Firstyear @ToxicMushroom 

image

Fixes #

Checklist

- [x] This PR contains no AI generated code
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

File: server/core/templates/base_htmx.html [hunk 1]
```diff
@@ -20,6 +20,8 @@
 		<script src="/pkg/external/bootstrap.bundle.min.js" integrity="sha384-YvpcrYf0tY3lHB60NNkmXc5s9fDVZLESaAA55NDzOxhy9GkcIdslK1eN7N6jIeHz"></script>
 		<script src="/pkg/external/htmx.min.1.9.12.js" integrity="sha384-ujb1lZYygJmzgSwoxRggbCHcjc0rB2XoQrxeTUQyRjrOnlCoYta87iKBWq3EsdM2"></script>
 		<link rel="stylesheet" href="/pkg/style.css" />
+		<script src="/pkg/external/cred_update.js"></script>
+		<script src="/pkg/external/base64.js" async></script>
```
Comment (ToxicMushroom) [inline server/core/templates/base_htmx.html:24]: These scripts should be imported in the places that require them, not everywhere

File: server/core/src/https/views/reset.rs [hunk 2]
```diff
@@ -744,21 +762,41 @@ fn get_cu_partial_response(cu_status: CUStatus) -> Response {
...
+    if is_logged_in {
+        let cred_status_view = CredStatusView {
+            uri,
             domain,
             names,
             credentials_update_partial,
-        }),
-    )
-        .into_response()
+            posix_enabled: false,
```
Comment (ToxicMushroom) [inline server/core/src/https/views/reset.rs:778]: add a TODO so I remember to fill this in :)

File: server/core/templates/user_settings_partial_base.html
Comment (ToxicMushroom) [inline server/core/templates/user_settings_partial_base.html:1]: I'm not a fan of how you used the uri/href matching here.
When you're writing an endpoint on the backend and preparing the response I think you always know what menu entry the user should be in.
So I'd pass a string or enum to activate 1 of the menu items.

**Review by ToxicMushroom — CHANGES_REQUESTED** [review]
Looks great, I'm happy with how it's looking so far.

I'll run it myself tomorrow to check further :)

File: server/core/src/https/views/reset.rs (hunk 2 above)
Comment (weijiangan) [reply server/core/src/https/views/reset.rs:778]: you mean like a comment above this right?

Comment (ToxicMushroom) [reply server/core/src/https/views/reset.rs:778]: yeah, // TODO: fill in posix enabled

File: server/core/templates/user_settings_partial_base.html [hunk 3]
```diff
@@ -1,10 +1,14 @@
+(% macro side_menu_item(label, href, uri, svg) %)
+<a hx-select="main" hx-target="main" hx-swap="outerHTML show:false swap:1s" href="(( href ))"
```
Comment (ToxicMushroom) [inline server/core/templates/user_settings_partial_base.html:2]: why is there a swap:1s here, this makes the swapping feel incredibly sluggish

File: server/core/src/https/views/reset.rs [hunk 4]
```diff
@@ -46,6 +52,16 @@ struct CredResetView {
     credentials_update_partial: CredResetPartialView,
 }
 
+#[derive(Template)]
+#[template(path = "credentials_status.html")]
+struct CredStatusView {
+    uri: Uri,
```
Comment (ToxicMushroom) [inline server/core/src/https/views/reset.rs:58]: I'd still prefer this be changed to an enum like I previously commented

File: server/core/src/https/views/reset.rs [hunk 5]
```diff
@@ -669,7 +687,7 @@ pub(crate) async fn view_reset_get(
         };
 
         // CU Session cookie is okay
-        let cu_resp = get_cu_response(domain_display_name, cu_status);
+        let cu_resp = get_cu_response(domain_display_name, cu_status, uri, false);
```
Comment (ToxicMushroom) [inline server/core/src/https/views/reset.rs:690]: this false is not correct, you should check 
```rust
    let logged_in = state
        .qe_r_ref
        .handle_auth_valid(client_auth_info.clone(), kopid.eventid).await.is_ok();
```

File: server/core/src/https/views/reset.rs [hunk 6]
```diff
@@ -679,7 +697,7 @@ pub(crate) async fn view_reset_get(
             .await
         {
             Ok((cu_session_token, cu_status)) => {
-                let cu_resp = get_cu_response(domain_display_name, cu_status);
+                let cu_resp = get_cu_response(domain_display_name, cu_status, uri, false);
```
Comment (ToxicMushroom) [inline server/core/src/https/views/reset.rs:700]: same here

**Review by ToxicMushroom — CHANGES_REQUESTED** [review]

File: server/core/templates/user_settings_partial_base.html [hunk 7]
```diff
@@ -1,10 +1,14 @@
...
+
 <main class="container-xxl pb-5">
     <div class="d-flex flex-sm-row flex-column">
-        <div class="list-group side-menu">
-            <a href="#" class="list-group-item list-group-item-action active">
-                <svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" fill="currentColor" class="bi bi-person-fill me-3" viewBox="0 0 16 16">
-                    <path d="M3 14s-1 0-1-1 1-4 6-4 6 3 6 4-1 1-1 1zm5-6a3 3 0 1 0 0-6 3 3 0 0 0 0 6"/>
-                </svg>Profile</a>
+        <div class="list-group side-menu flex-shrink-0">
+            (% call side_menu_item("Profile", "/ui/profile", uri, "<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"20\" height=\"20\" fill=\"currentColor\" class=\"bi bi-person-fill me-3\" viewBox=\"0 0 16 16\"><path d=\"M3 14s-1 0-1-1 1-4 6-4 6 3 6 4-1 1-1 1zm5-6a3 3 0 1 0 0-6 3 3 0 0 0 0 6\"/></svg>") %)
```
Comment (yaleman) [inline server/core/templates/user_settings_partial_base.html:11]: Seems like someone's already been doing it, but please don't include inline SVG like this, please make them static files. We also need to be able to identify where they came from for licensing purposes.

**Review by yaleman — CHANGES_REQUESTED** [review]
Per comment, don't include SVG inline.

File: server/core/templates/user_settings_partial_base.html (hunk 7 above)
Comment (ToxicMushroom) [reply server/core/templates/user_settings_partial_base.html:11]: This was my doings, they're bootstrap icons under MIT https://github.com/twbs/icons

**Review by ToxicMushroom — APPROVED** [review]
looks good

**Comment by Firstyear** [comment]
Sorry for my relative silence, I'm going to try to review this today :)

**Review by Firstyear — DISMISSED** [review]
@weijiangan I rebased for you.

I gave this a test and it works well. It'd be cool to re-add the "add another device" flow too from the wasm. It allows the user to self-create their own reset tokens and qr codes so they can scan the qr code with a phone or similar to enroll that device :)

**Comment by weijiangan** [comment]
@Firstyear I'll add it back :)

File: server/core/src/https/views/mod.rs [hunk 8]
```diff
@@ -36,7 +37,7 @@ pub fn view_router() -> Router<ServerState> {
         .route("/", get(|| async { Redirect::permanent("/ui/login") }))
         .route("/apps", get(apps::view_apps_get))
         .route("/reset", get(reset::view_reset_get))
-        .route("/update_credentials", get(reset::view_self_reset_get))
+        .route("/update-credentials", get(reset::view_self_reset_get))
```
Comment (yaleman) [inline server/core/src/https/views/mod.rs:40]: Why was this changed? We use underscores in URLs.

```suggestion
        .route("/update_credentials", get(reset::view_self_reset_get))
```

File: server/core/src/https/views/reset.rs [hunk 9]
```diff
@@ -600,7 +617,7 @@ pub(crate) async fn view_self_reset_get(
             client_auth_info,
             kopid,
             jar,
-            "/ui/update_credentials",
+            "/ui/update-credentials",
```
Comment (yaleman) [inline server/core/src/https/views/reset.rs:620]: As per previous comment

```suggestion
            "/ui/update_credentials",
```

File: server/core/src/https/views/reset.rs [hunk 10]
```diff
@@ -736,21 +759,46 @@ fn get_cu_partial_response(cu_status: CUStatus) -> Response {
...
-        .into_response()
+            // TODO: fill in posix enabled
+            posix_enabled: false,
+        };
+        let profile_view = ProfileView {
+            profile_partial: cred_status_view,
+        };
+
+        (
+            HxPushUrl(Uri::from_static("/ui/update-credentials")),
```
Comment (yaleman) [inline server/core/src/https/views/reset.rs:787]:
```suggestion
            HxPushUrl(Uri::from_static("/ui/update_credentials")),
```

File: server/core/templates/user_settings_partial_base.html [hunk 11]
```diff
@@ -1,24 +1,19 @@
...
-                </svg>UNIX Password</a>
+                (% call side_menu_item("UNIX Password", "/ui/update-credentials", ProfileMenuItems::UnixPassword, "building-lock") %)
             (% endif %)
-
-            <a href="/ui/update_credentials" class="list-group-item list-group-item-action"><svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" fill="currentColor" class="bi bi-shield-fill-plus me-3" viewBox="0 0 16 16">
-                <path fill-rule="evenodd" d="M8 0c-.69 0-1.843.265-2.928.56-1.11.3-2.229.655-2.887.87a1.54 1.54 0 0 0-1.044 1.262c-.596 4.477.787 7.795 2.465 9.99a11.8 11.8 0 0 0 2.517 2.453c.386.273.744.482 1.048.625.28.132.581.24.829.24s.548-.108.829-.24a7 7 0 0 0 1.048-.625 11.8 11.8 0 0 0 2.517-2.453c1.678-2.195 3.061-5.513 2.465-9.99a1.54 1.54 0 0 0-1.044-1.263 63 63 0 0 0-2.887-.87C9.843.266 8.69 0 8 0m-.5 5a.5.5 0 0 1 1 0v1.5H10a.5.5 0 0 1 0 1H8.5V9a.5.5 0 0 1-1 0V7.5H6a.5.5 0 0 1 0-1h1.5z"/>
-            </svg>Add another device</a>
+            (% call side_menu_item("Credentials", "/ui/update-credentials", ProfileMenuItems::Credentials, "shield-lock") %)
```
Comment (yaleman) [inline server/core/templates/user_settings_partial_base.html:16]:
```suggestion
            (% call side_menu_item("Credentials", "/ui/update_credentials", ProfileMenuItems::Credentials, "shield-lock") %)
```

File: server/core/templates/user_settings_partial_base.html [hunk 12]
```diff
@@ -1,24 +1,19 @@
...
-                    <path d="M2 1a1 1 0 0 1 1-1h10a1 1 0 0 1 1 1v7.764a3 3 0 0 0-4.989 2.497 2 2 0 0 0-.743.739H6.5a.5.5 0 0 0-.5.5V16H3a1 1 0 0 1-1-1zm2 1.5v1a.5.5 0 0 0 .5.5h1a.5.5 0 0 0 .5-.5v-1a.5.5 0 0 0-.5-.5h-1a.5.5 0 0 0-.5.5m3 0v1a.5.5 0 0 0 .5.5h1a.5.5 0 0 0 .5-.5v-1a.5.5 0 0 0-.5-.5h-1a.5.5 0 0 0-.5.5m3.5-.5a.5.5 0 0 0-.5.5v1a.5.5 0 0 0 .5.5h1a.5.5 0 0 0 .5-.5v-1a.5.5 0 0 0-.5-.5zM4 5.5v1a.5.5 0 0 0 .5.5h1a.5.5 0 0 0 .5-.5v-1a.5.5 0 0 0-.5-.5h-1a.5.5 0 0 0-.5.5M7.5 5a.5.5 0 0 0-.5.5v1a.5.5 0 0 0 .5.5h1a.5.5 0 0 0 .5-.5v-1a.5.5 0 0 0-.5-.5zm2.5.5v1a.5.5 0 0 0 .5.5h1a.5.5 0 0 0 .5-.5v-1a.5.5 0 0 0-.5-.5h-1a.5.5 0 0 0-.5.5M4.5 8a.5.5 0 0 0-.5.5v1a.5.5 0 0 0 .5.5h1a.5.5 0 0 0 .5-.5v-1a.5.5 0 0 0-.5-.5zm2.5.5v1a.5.5 0 0 0 .5.5h1a.5.5 0 0 0 .5-.5v-1a.5.5 0 0 0-.5-.5h-1a.5.5 0 0 0-.5.5"/>
-                    <path d="M9 13a1 1 0 0 1 1-1v-1a2 2 0 1 1 4 0v1a1 1 0 0 1 1 1v2a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1zm3-3a1 1 0 0 0-1 1v1h2v-1a1 1 0 0 0-1-1"/>
-                </svg>UNIX Password</a>
+                (% call side_menu_item("UNIX Password", "/ui/update-credentials", ProfileMenuItems::UnixPassword, "building-lock") %)
```
Comment (yaleman) [inline server/core/templates/user_settings_partial_base.html:14]:
```suggestion
                (% call side_menu_item("UNIX Password", "/ui/update_credentials", ProfileMenuItems::UnixPassword, "building-lock") %)
```

File: server/core/src/https/views/constants.rs [hunk 13]
```diff
@@ -0,0 +1,7 @@
+#[derive(PartialEq, Clone)]
```
Comment (yaleman) [inline server/core/src/https/views/constants.rs:1]: This should have a serde::{Serialize,Deserialize} with the container attribute `rename_all="snake_case"` to allow for usage/parsing in URLs and other places

**Review by yaleman — CHANGES_REQUESTED** [review]
Looking good so far

**Review by yaleman — APPROVED** [review]

**Review by Firstyear — APPROVED** [review]
